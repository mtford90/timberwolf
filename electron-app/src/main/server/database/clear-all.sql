-- noinspection SqlWithoutWhereForFile

DELETE FROM logs;
DELETE FROM words;
DELETE FROM sources;
