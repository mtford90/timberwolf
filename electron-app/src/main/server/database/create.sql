CREATE TABLE IF NOT EXISTS sources
(
    id            integer PRIMARY KEY,
    name          text NOT NULL
);

CREATE UNIQUE INDEX index_source_name
    ON sources (name);

CREATE TABLE IF NOT EXISTS logs
(
    source_id integer NOT NULL,
    timestamp integer NOT NULL,
    text      text    NOT NULL,
    FOREIGN KEY (source_id) REFERENCES sources (id)
        ON DELETE CASCADE ON UPDATE NO ACTION
);

CREATE INDEX IF NOT EXISTS idx_timestamp
    ON logs (timestamp);

CREATE TABLE IF NOT EXISTS words
(
    source_id integer NOT NULL,
    text      text    NOT NULL,
    num       integer NOT NULL,
    UNIQUE (source_id, text),
    FOREIGN KEY (source_id) REFERENCES sources (id)
        ON DELETE CASCADE ON UPDATE NO ACTION
);

CREATE INDEX IF NOT EXISTS idx_words_source
    ON words (source_id);

CREATE INDEX IF NOT EXISTS idx_words_text
    ON words (text);

CREATE INDEX IF NOT EXISTS idx_words_num
    ON words (num);
