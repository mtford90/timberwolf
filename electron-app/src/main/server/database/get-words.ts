import { compact, flatten, groupBy, keys } from "lodash";

export type WordEntries = [number, string[]][];

export function getWords(
  rows: Array<{ sourceId: number; text: string }>
): WordEntries {
  const grouped = groupBy(
    rows.map((r) => {
      const withoutSymbols = r.text.replace(
        /[-!$%^&*()_+|~=`{}[\]:";'<>?,./]/g,
        " "
      );

      return {
        sourceId: r.sourceId,
        words: [
          ...withoutSymbols.split(/\s/g),
          ...r.text.split(/(?!\(.*)\s(?![^(]*?\))/g),
        ],
      };
    }),
    (r) => r.sourceId
  );

  const entries: WordEntries = [];

  keys(grouped).forEach((source) => {
    const words = compact(flatten(grouped[source].map((g) => g.words)));
    entries.push([parseInt(source, 10), words]);
  });

  return entries;
}
