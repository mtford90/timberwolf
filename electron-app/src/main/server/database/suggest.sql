SELECT *
FROM words
WHERE text LIKE @query
  AND source_id = @sourceId
ORDER BY num DESC, length(text)
LIMIT @limit OFFSET @offset;
