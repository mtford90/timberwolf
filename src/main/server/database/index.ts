import sqlite, { Database as SqliteDatabase } from "better-sqlite3";

export class Database {
  private db: SqliteDatabase;

  constructor(path = ":memory:") {
    this.db = sqlite(path);
  }

  init() {
    this.db.exec(
      `
      CREATE TABLE IF NOT EXISTS logs
      (
          path      text NOT NULL,
          timestamp integer NOT NULL,
          text      text NOT NULL
      );
      
      CREATE INDEX IF NOT EXISTS idx_path
      ON logs(path);
      
      CREATE INDEX IF NOT EXISTS idx_timestamp
      ON logs(timestamp);
    `
    );
  }

  insert(rows: Array<{ path: string; timestamp?: number; text: string }>) {
    const stmt = this.db.prepare(
      "INSERT INTO logs (path, timestamp, text) VALUES (?,?,?)"
    );
    rows.forEach(({ path, timestamp, text }) => {
      stmt.run(path, timestamp || Date.now(), text);
    });
  }

  lines(
    path: string,
    text: string,
    opts: {
      offset?: number;
      limit?: number;
      fields?: Array<"rowid" | "path" | "timestamp" | "text">;
    } = {}
  ): Promise<
    Array<{ rowid: string; path: string; timestamp: number; text: string }>
  > {
    const offset = opts.offset || 0;
    const limit = opts.limit || 10;
    const fields = opts.fields || ["rowid", "path", "timestamp", "text"];

    const query = `
      SELECT ${fields.join(",")}
      FROM logs
      WHERE text LIKE '%${text}%'
      AND path = '${path}'
      ORDER BY rowid asc
      LIMIT  ${limit}
      OFFSET ${offset};
    `;

    return this.db.prepare(query).all() as any;
  }

  numLines(path: string) {
    return this.db
      .prepare(`SELECT COUNT(*) as n FROM logs WHERE path = '${path}'`)
      .get().n;
  }

  clear(path: string) {
    this.db.exec(`DELETE FROM logs WHERE path = '${path}'`);
  }

  clearAll() {
    this.db.exec("DELETE FROM logs");
  }

  close() {
    this.db.close();
  }
}
