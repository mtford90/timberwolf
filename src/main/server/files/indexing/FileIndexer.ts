import mitt, { Emitter } from "mitt";
import { Subscription } from "rxjs";
import { Database } from "../../database";
import { reverseRead } from "./reverse-read";

export type FileIndexerProgressEvent = {
  path: string;
  bytesRead: number;
  totalBytes: number;
};

export type FileIndexerErrorEvent = { path: string; error: Error };

export type FileIndexerCloseEvent = { path: string };

export type FileIndexerEventPayloads = {
  progress: FileIndexerProgressEvent;
  close: FileIndexerCloseEvent;
  error: FileIndexerErrorEvent;
};

export type FileIndexerEvent = keyof FileIndexerEventPayloads;

export class FileIndexer {
  private database: Database;

  private emitter: Emitter;

  public files: Map<
    string,
    {
      lines: number;
    }
  > = new Map();

  public subscriptions: Map<string, Subscription> = new Map();

  constructor(database: Database) {
    this.database = database;
    this.emitter = mitt();
  }

  on<T extends FileIndexerEvent>(
    name: T,
    fn: (payload: FileIndexerEventPayloads[T]) => void
  ) {
    this.emitter.on<FileIndexerEventPayloads[T]>(name, fn as never);
    return () => this.off(name, fn);
  }

  off<T extends FileIndexerEvent>(
    name: T,
    fn: (payload: FileIndexerEventPayloads[T]) => void
  ) {
    this.emitter.off<FileIndexerEventPayloads[T]>(name, fn as never);
  }

  protected emit<T extends FileIndexerEvent>(
    name: T,
    payload: FileIndexerEventPayloads[T]
  ) {
    this.emitter.emit<FileIndexerEventPayloads[T]>(name, payload);
  }

  index(path: string, encoding: BufferEncoding = "utf8") {
    if (this.subscriptions.has(path)) {
      // no-op - already indexing the file
      return;
    }

    const observable = reverseRead(path, encoding);

    const subscription = observable.subscribe({
      next: (o) => {
        if (o.line === undefined) {
          this.files.set(path, { lines: 0 });
        }
        this.database.insert([{ path, text: o.line || "\n" }]);
        this.emit("progress", {
          totalBytes: o.size,
          path,
          bytesRead: o.bytesRead,
        });
        const fileData = this.files.get(path);
        if (fileData) {
          const newLines = fileData.lines + 1;
          this.files.set(path, {
            ...fileData,
            lines: newLines,
          });
        }
      },
      error: (err) => {
        this.emit("error", { error: err, path });
      },
      complete: () => {
        console.log("complete");
        this.emit("close", { path });
      },
    });

    this.subscriptions.set(path, subscription);
  }

  cancel(path: string) {
    const sub = this.subscriptions.get(path);
    if (sub) {
      sub.unsubscribe();
      this.subscriptions.delete(path);
    }
  }
}
