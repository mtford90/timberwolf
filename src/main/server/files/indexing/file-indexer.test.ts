import { last } from "lodash";
import { FileIndexer, FileIndexerProgressEvent } from "./FileIndexer";
import { Database } from "../../database";

describe("Indexer", () => {
  let db: Database;
  let indexer: FileIndexer;
  const path = "/Users/mtford/Playground/log/log.txt";

  beforeEach(() => {
    db = new Database();
    indexer = new FileIndexer(db);
    return db.init();
  });

  describe("indexing", () => {
    it("should index the lines", async () => {
      const results = await db.lines(path, "workflow-upload-orchestrator", {
        limit: 100,
      });

      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(results).toMatchSnapshot();

      results.forEach((result) => {
        expect(result.text).toContain("workflow-upload-orchestrator");
      });
    });

    describe("events", () => {
      let progressEvents: FileIndexerProgressEvent[] = [];

      beforeEach((done) => {
        const subscriptions = [
          indexer.on("close", () => {
            console.log("closed!");
            subscriptions.forEach((fn) => fn());
            done();
          }),
          indexer.on("progress", (e) => progressEvents.push(e)),
          indexer.on("error", done),
        ];

        indexer.index(path);
      });

      afterEach(() => {
        progressEvents = [];
      });

      it("should emit progress events whilst indexing the file", async () => {
        expect(progressEvents.length).toBeTruthy();
        const lastEvent = last(progressEvents)!;
        expect(lastEvent.bytesRead / lastEvent.totalBytes).toEqual(1);
      });
    });

    describe("cache", () => {
      describe("when indexing has finished", () => {
        beforeEach((done) => {
          indexer.index(path);
          indexer.on("close", () => done());
        });

        it("should have cached the number of lines", async () => {
          expect(indexer.files.get(path)?.lines).toEqual(74);
        });
      });
    });
  });
});
