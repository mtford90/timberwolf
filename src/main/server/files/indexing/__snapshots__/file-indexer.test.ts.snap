// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indexer indexing should index the lines 1`] = `Array []`;

exports[`Indexer should index the lines 1`] = `
Array [
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 2,
    "text": " DEBUG  [workflow-upload-orchestrator] no operations performed - sleeping",
    "timestamp": 1597250458409,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 3,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458411,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 4,
    "text": " DEBUG  [workflow-upload-orchestrator] event was received during previous operation - trying another tick",
    "timestamp": 1597250458413,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 5,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
    "timestamp": 1597250458416,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 6,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/error_1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073741952, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
    "timestamp": 1597250458427,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 7,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
    "timestamp": 1597250458430,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 8,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073741888, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [64]}",
    "timestamp": 1597250458443,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 9,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458444,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 10,
    "text": " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
    "timestamp": 1597250458446,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 11,
    "text": " DEBUG  [workflow-upload-orchestrator] marked workflow folder as errored",
    "timestamp": 1597250458447,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 12,
    "text": " DEBUG  [workflow-upload-orchestrator] no payload file to upload",
    "timestamp": 1597250458449,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 13,
    "text": " DEBUG  [workflow-upload-orchestrator] no binary info files to upload",
    "timestamp": 1597250458451,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 14,
    "text": " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
    "timestamp": 1597250458453,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 15,
    "text": " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": null, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
    "timestamp": 1597250458468,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 16,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
    "timestamp": 1597250458470,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 17,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/workflow_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 512, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [512]}",
    "timestamp": 1597250458482,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 18,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458483,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 19,
    "text": " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
    "timestamp": 1597250458484,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 20,
    "text": " DEBUG  [workflow-upload-orchestrator] deleted workflow payload",
    "timestamp": 1597250458485,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 23,
    "text": " DEBUG  [workflow-upload-orchestrator] no binary info files to upload",
    "timestamp": 1597250458488,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 24,
    "text": " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
    "timestamp": 1597250458488,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 25,
    "text": " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
    "timestamp": 1597250458505,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 26,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458505,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 27,
    "text": " DEBUG  [workflow-upload-orchestrator] error when performing operation - sleeping for 250ms before ticking again",
    "timestamp": 1597250458507,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 28,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
    "timestamp": 1597250458508,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 29,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/error_info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
    "timestamp": 1597250458515,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 30,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
    "timestamp": 1597250458516,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 31,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 64, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [64]}",
    "timestamp": 1597250458523,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 32,
    "text": " DEBUG  [workflow-upload-orchestrator] marking session binary info as errored",
    "timestamp": 1597250458524,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 33,
    "text": " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
    "timestamp": 1597250458525,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 34,
    "text": " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
    "timestamp": 1597250458540,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 35,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458540,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 36,
    "text": " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
    "timestamp": 1597250458541,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 37,
    "text": " DEBUG  [workflow-upload-orchestrator] deleted session binary",
    "timestamp": 1597250458542,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 38,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (4 ignored events)",
    "timestamp": 1597250458544,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 39,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 512, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [512]}",
    "timestamp": 1597250458552,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 42,
    "text": " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
    "timestamp": 1597250458580,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 43,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (3 ignored events)",
    "timestamp": 1597250458581,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 44,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/workflow_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 256, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [256]}",
    "timestamp": 1597250458589,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 45,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
    "timestamp": 1597250458591,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 46,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
    "timestamp": 1597250458598,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 47,
    "text": " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
    "timestamp": 1597250458599,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 48,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
    "timestamp": 1597250458606,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 49,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458607,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 50,
    "text": " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073742080, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [256]}",
    "timestamp": 1597250458615,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 52,
    "text": " DEBUG  [workflow-upload-orchestrator] no operations performed - sleeping",
    "timestamp": 1597250458621,
  },
  Object {
    "path": "/Users/mtford/Playground/log/log.txt",
    "rowid": 53,
    "text": " DEBUG  [workflow-upload-orchestrator] performOperation",
    "timestamp": 1597250458621,
  },
]
`;
